### PR TITLE
Exit process with code 1 when outdated dependencies are found

### LINKED ISSUE #3483: Make "yarn outdated" exit with a non-zero exit code when there are outdated dependencies
**Do you want to request a *feature* or report a *bug*?**

Feature (probably).

**What is the current behavior?**

With outdated packages present, `yarn outdated` exits with `0`

The `npm outdated ` in the same situation will exit with `1` (as of NPM `4.2.0`).

**What is the expected behavior?**

With outdated packages present, `yarn outdated` should exit with non-zero exit code.

**Please mention your node.js, yarn and operating system version.**

OS: Debian Jessie inside the Docker, [`library/node` image](https://hub.docker.com/_/node/), `7.10` tag.
Node: `v7.10.0`
Yarn: `0.24.4`

**P.S.**
Bundlers `bundle outdated` also returns non-zero exit code if outdated gems are present, just for the reference.
We use that part of behavior of both Bundler and NPM in automated checks on CI.

### PR DESCRIPTION
**Pull request by KishanBagaria**
**Summary**
Fixes #3483 

**Test plan**

### CODE REVIEW

File: src/cli/commands/outdated.js [hunk 1]
```diff
@@ -44,6 +44,9 @@ export async function run(config: Config, reporter: Reporter, flags: Object, arg
     });
 
     reporter.table(['Package', 'Current', 'Wanted', 'Latest', 'Package Type', 'URL'], body);
+    process.on('exit', () =>
+      process.exitCode = 1
```
Comment (bestander) [inline src/cli/commands/outdated.js]: That looks like a hack.
We prefer to have a single exit code in cli/index.js that will be set code 1 if the command returned a rejected promise

**Comment by KishanBagaria** [comment]
Good idea. What do you think of the update?

This is an alternative way of setting the exit code:
```
process.on('exit', () =>
  process.exit(EXIT_CODE)
);
```

[NPM does it with `process.exitCode = EXIT_CODE`](https://github.com/npm/npm/blob/latest/lib/outdated.js#L115) but we can't do that because of commander (see first comment above).

File: src/cli/index.js [hunk 2]
```diff
@@ -155,11 +155,16 @@ if (command.requireLockfile && !fs.existsSync(path.join(config.cwd, constants.LO
 //
 const run = (): Promise<void> => {
   invariant(command, 'missing command');
-  return command.run(config, reporter, commander, commander.args).then(() => {
+  return command.run(config, reporter, commander, commander.args).then(exitCode => {
     reporter.close();
     if (outputWrapper) {
       reporter.footer(false);
     }
```
Comment (bestander) [inline src/cli/index.js]: That would not work well, commands may resolve with non numeric values.
Also at line 337 in this file a rejection of run() promise causes process.exit(1) already.

Can you just throw an error in the outdated command?

Comment (KishanBagaria) [inline src/cli/index.js]: Throwing an error causes the `yarn-error.log` file to be written and error message stuff to be printed. Am I missing something here?

I did a find all and none of the commands return anything. They all have the signature of `Promise<void>`. I've updated the PR again.

**Comment by KishanBagaria** [comment]
It wasn't commander that was preventing `process.exitCode = EXIT_CODE` from working. It was yarn's `exit` method.

Another alternative way to do this:
The `exit` method in `cli/index.js` can be modified to not pass `0` to `process.exit`, and `outdated.js` can mutate `process.exitCode = 1`.

**Comment by bestander** [comment]
I think this is better than modifying process, thanks @KishanBagaria!

**Pull request merged**